File: c/src/matrix_mul.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"

typedef struct
{
    long start_row;
    long end_row;
    long n;
    const double *a;
    const double *b_transposed;
    double *c;
} worker_args_t;

static double value_for(long row, long col)
{
    return (double)(((row + 1) * 31 + (col + 1) * 17) % 101) / 101.0;
}

static void fill_matrices(double *a, double *b, double *b_transposed, long n)
{
    for (long row = 0; row < n; row++)
    {
        for (long col = 0; col < n; col++)
        {
            a[row * n + col] = value_for(row, col);
            b[row * n + col] = value_for(col, row);
            b_transposed[col * n + row] = b[row * n + col];
        }
    }
}

static void *worker(void *arg)
{
    worker_args_t *args = (worker_args_t *)arg;
    long n = args->n;

    for (long row = args->start_row; row < args->end_row; row++)
    {
        for (long col = 0; col < n; col++)
        {
            double sum = 0.0;
            const double *a_row = &args->a[row * n];
            const double *b_col = &args->b_transposed[col * n];

            for (long k = 0; k < n; k++)
            {
                sum += a_row[k] * b_col[k];
            }

            args->c[row * n + col] = sum;
        }
    }

    return NULL;
}

int main(int argc, char **argv)
{
    long threads;
    long n;
    size_t elements;
    double *a;
    double *b;
    double *b_transposed;
    double *c;
    pthread_t *thread_ids;
    worker_args_t *args;
    double start;
    double elapsed;
    double checksum = 0.0;
    char extra[80];
    long rows_per_thread;

    if (argc != 3)
    {
        fprintf(stderr, "Usage: %s <threads> <matrix_size>\n", argv[0]);
        return EXIT_FAILURE;
    }

    threads = parse_positive_long(argv[1], "threads");
    n = parse_positive_long(argv[2], "matrix_size");
    elements = (size_t)n * (size_t)n;

    a = malloc(elements * sizeof(*a));
    b = malloc(elements * sizeof(*b));
    b_transposed = malloc(elements * sizeof(*b_transposed));
    c = calloc(elements, sizeof(*c));
    thread_ids = calloc((size_t)threads, sizeof(*thread_ids));
    args = calloc((size_t)threads, sizeof(*args));

    if (a == NULL || b == NULL || b_transposed == NULL || c == NULL ||
        thread_ids == NULL || args == NULL)
    {
        perror("allocation");
        free(a);
        free(b);
        free(b_transposed);
        free(c);
        free(thread_ids);
        free(args);
        return EXIT_FAILURE;
    }

    fill_matrices(a, b, b_transposed, n);
    rows_per_thread = (n + threads - 1) / threads;

    start = monotonic_seconds();

    for (long i = 0; i < threads; i++)
    {
        int error_number;
        long start_row = i * rows_per_thread;
        long end_row = start_row + rows_per_thread;

        if (end_row > n)
        {
            end_row = n;
        }

        args[i].start_row = start_row;
        args[i].end_row = end_row;
        args[i].n = n;
        args[i].a = a;
        args[i].b_transposed = b_transposed;
        args[i].c = c;

        error_number = pthread_create(&thread_ids[i], NULL, worker, &args[i]);
        if (error_number != 0)
        {
            fprintf(stderr, "pthread_create: %s\n", strerror(error_number));
            free(a);
            free(b);
            free(b_transposed);
            free(c);
            free(thread_ids);
            free(args);
            return EXIT_FAILURE;
        }
    }

    for (long i = 0; i < threads; i++)
    {
        int error_number = pthread_join(thread_ids[i], NULL);
        if (error_number != 0)
        {
            fprintf(stderr, "pthread_join: %s\n", strerror(error_number));
            free(a);
            free(b);
            free(b_transposed);
            free(c);
            free(thread_ids);
            free(args);
            return EXIT_FAILURE;
        }
    }

    elapsed = monotonic_seconds() - start;

    for (size_t i = 0; i < elements; i++)
    {
        checksum += c[i];
    }

    snprintf(extra, sizeof(extra), "checksum=%.6f", checksum);
    print_csv_result("c", "matrix_mul", threads, n, elapsed, extra);

    free(a);
    free(b);
    free(b_transposed);
    free(c);
    free(thread_ids);
    free(args);
    return EXIT_SUCCESS;
}

File: rust/src/bin/matrix_mul.rs
use std::env;
use std::process;
use std::thread;
use std::time::Instant;

fn parse_positive(value: &str, name: &str) -> usize {
    match value.parse::<usize>() {
        Ok(parsed) if parsed > 0 => parsed,
        _ => {
            eprintln!("Invalid {name}: {value}");
            process::exit(1);
        }
    }
}

fn value_for(row: usize, col: usize) -> f64 {
    (((row + 1) * 31 + (col + 1) * 17) % 101) as f64 / 101.0
}

fn fill_matrices(n: usize) -> (Vec<f64>, Vec<f64>) {
    let mut a = vec![0.0; n * n];
    let mut b_transposed = vec![0.0; n * n];

    for row in 0..n {
        for col in 0..n {
            a[row * n + col] = value_for(row, col);
            b_transposed[col * n + row] = value_for(col, row);
        }
    }

    (a, b_transposed)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} <threads> <matrix_size>", args[0]);
        process::exit(1);
    }

    let threads = parse_positive(&args[1], "threads");
    let n = parse_positive(&args[2], "matrix_size");
    let (a, b_transposed) = fill_matrices(n);
    let mut c = vec![0.0; n * n];
    let rows_per_thread = (n + threads - 1) / threads;

    let start = Instant::now();

    thread::scope(|scope| {
        for (chunk_id, c_chunk) in c.chunks_mut(rows_per_thread * n).enumerate() {
            let start_row = chunk_id * rows_per_thread;
            let a = &a;
            let b_transposed = &b_transposed;

            scope.spawn(move || {
                let rows_in_chunk = c_chunk.len() / n;

                for local_row in 0..rows_in_chunk {
                    let row = start_row + local_row;
                    let output_row = &mut c_chunk[local_row * n..(local_row + 1) * n];

                    for col in 0..n {
                        let a_row = &a[row * n..(row + 1) * n];
                        let b_col = &b_transposed[col * n..(col + 1) * n];
                        let mut sum = 0.0;

                        for k in 0..n {
                            sum += a_row[k] * b_col[k];
                        }

                        output_row[col] = sum;
                    }
                }
            });
        }
    });

    let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
    let checksum: f64 = c.iter().sum();
    println!("rust,matrix_mul,{threads},{n},{elapsed_ms:.6},checksum={checksum:.6}");
}
